### PR TITLE
feat: 109 isometric building workflow

### PR DESCRIPTION
**Pull request by TheLLspectre**
Adding isometric workflow:
Main features: 

- Select Base and also custom base
- select model
- define and select prompt to generate assets
- Select generated asset to convert as sprite, as a tile to tile palette, customize in webapp
- Regenerate an inference
- Restart workflow

 - closes: #109

### CODE REVIEW

File: package/Editor/Images/Images.cs [hunk 1]
```diff
@@ -162,5 +163,91 @@ public static Texture2D GetTextureByImageId(string _id)
...
+
+            _onAction?.Invoke();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="_texture"></param>
+        /// <param name="_onAction"></param>
+        public static void DlAsTexture(Texture2D _texture, Action _onAction)
```
Comment (Morgan-6Freedom) [inline package/Editor/Images/Images.cs:185]: I prefer full wording "DownloadAsTexture" + it miss the summary

File: package/Editor/Images/Images.cs [hunk 2]
```diff
@@ -162,5 +163,91 @@ public static Texture2D GetTextureByImageId(string _id)
...
+            _onAction?.Invoke();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="_texture"></param>
+        /// <param name="_selectedImageId"></param>
+        /// <param name="_onAction"></param>
+        public static void DlAsSprite(Texture2D _texture, string _selectedImageId, Action _onSuccessAction, Action _onDownloadingAction)
```
Comment (Morgan-6Freedom) [inline package/Editor/Images/Images.cs:197]: same as "DlAsTexture"

File: package/Editor/Images/Images.cs [hunk 3]
```diff
@@ -162,5 +163,91 @@ public static Texture2D GetTextureByImageId(string _id)
             return imageData.texture;
         }
 
+        /// <summary>
+        ///
```
Comment (Morgan-6Freedom) [inline package/Editor/Images/Images.cs:167]: it miss the summary here

File: package/Editor/Isometric Workflow/IsometricWorkflow.cs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        }
+
+        private void CreateGUI()
+        {
+            var isometricWorkflow = (IsometricWorkflow)GetWindow(typeof(IsometricWorkflow));
+            isometricWorkflowUI.Init(isometricWorkflow);
+        }
+
+        /// <summary>
+        /// Call the prompt window to generate one image per asset in the assetlist
```
Comment (Morgan-6Freedom) [inline package/Editor/Isometric Workflow/IsometricWorkflow.cs:120]: it's not one image, but 4. the comment is incorrect

File: package/Editor/Isometric Workflow/IsometricWorkflow.cs [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+            int totalRequests = assetList.Count;
+            int completedRequests = 0;
+
+            foreach (string assetName in assetList)
+            {
+                string tempName = assetName;
+                bool useBaseTexture = baseTexture != null;
+                string modelName = settings.isometricModels.Find(x => x.style == selectedModel).modelData.name;
+                PromptWindow.GenerateImage(modelName, useBaseTexture, useBaseTexture, useBaseTexture ? baseTexture : null, 4, $"isometric, solo, centered, solid color background, {selectedTheme}, {assetName}", 30, 1024, 1024, 6, "-1", useBaseTexture, 0.8f,
```
Comment (Morgan-6Freedom) [inline package/Editor/Isometric Workflow/IsometricWorkflow.cs:151]: I think we should make the prompt value "isometric, solo, centered, solid color background" as a variable so we can change it easily

File: package/Editor/Isometric Workflow/IsometricWorkflow.cs [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                    break;
+            }
+
+            int totalRequests = assetList.Count;
+            int completedRequests = 0;
+
+            string tempName = _assetName;
+            bool useBaseTexture = baseTexture != null;
+            string modelName = settings.isometricModels.Find(x => x.style == selectedModel).modelData.name;
+            PromptWindow.GenerateImage(modelName, useBaseTexture, useBaseTexture, useBaseTexture ? baseTexture : null, 4, $"isometric, solo, centered, solid color background, {selectedTheme}, {_assetName}", 30, 1024, 1024, 6, "-1", useBaseTexture, 0.8f,
```
Comment (Morgan-6Freedom) [inline package/Editor/Isometric Workflow/IsometricWorkflow.cs:201]: ⚠ The prompt is repeted here

File: package/Editor/Isometric Workflow/IsometricWorkflowSettings.cs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+            public ModelIdByStyle(IsometricWorkflow.ModelStyle style, string id)
+            {
+                this.style = style;
+                this.id = id;
+            }
+        }
+
+
+        private static IsometricWorkflowSettings GetOrCreateSettings()
```
Comment (Morgan-6Freedom) [inline package/Editor/Isometric Workflow/IsometricWorkflowSettings.cs:73]: Miss a summary here. Explain how the settings are made and what's stored inside

File: package/Editor/Isometric Workflow/IsometricWorkflowSettings.cs [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+            return false;
+        }
+
+
+        private void InitializeTextures()
+        {
+            squareBaseTexture = AssetDatabase.LoadAssetAtPath<Texture2D>(Path.Combine(CommonUtils.PluginFolderPath(), "Assets", "Reference Images", "IsometricBase_Square.png"));
+        }
+
+        private async void InitializeModels()
```
Comment (Morgan-6Freedom) [inline package/Editor/Isometric Workflow/IsometricWorkflowSettings.cs:133]: Miss a summary here to explain how the models are downloaded & stored

File: package/Editor/PromptImages/TileCreator.cs [hunk 9]
```diff
@@ -9,6 +9,14 @@ namespace Scenario.Editor
 {
     public class TileCreator
     {
+        #region Public Fields
+
+        public static TileCreator instance = null;
```
Comment (Morgan-6Freedom) [inline package/Editor/PromptImages/TileCreator.cs:14]: can you call this variable "Instance" with a capitalized i ?

File: package/Editor/PromptWindow/PromptWindow.cs [hunk 10]
```diff
@@ -1,3 +1,4 @@
+using Codice.CM.Common;
```
Comment (Morgan-6Freedom) [inline package/Editor/PromptWindow/PromptWindow.cs:1]: What is this namespace for ?

File: package/Editor/PromptWindow/PromptWindowUI.cs [hunk 11]
```diff
@@ -137,13 +137,15 @@ public void Render(Rect position)
...
 
-            CustomStyle.ButtonPrimary("Generate Image", 40, () =>
+            CustomStyle.ButtonPrimary("Generate Image", 40, 0, () =>
             {
                 promptinputText = SerializeTags(tags);
                 negativepromptinputText = SerializeTags(negativeTags);
 
                 EditorPrefs.SetString("postedModelName", DataCache.instance.SelectedModelId);
 
+                PromptFetcher.IsWorkflow = false;
```
Comment (Morgan-6Freedom) [inline package/Editor/PromptWindow/PromptWindowUI.cs:147]: What is this used for ?

File: package/Editor/Settings/PluginSettings.cs [hunk 12]
```diff
@@ -12,21 +12,32 @@ public class PluginSettings : EditorWindow
...
         public static string ApiUrl => "https://api.cloud.scenario.com/v1";
+        public static string WebAppUrl { get { return webAppUrl; } }
         public static Preset TexturePreset { get { return GetPreset(EditorPrefs.GetString("scenario/texturePreset")); } }
         public static Preset SpritePreset { get { return GetPreset(EditorPrefs.GetString("scenario/spritePreset")); } }
         public static Preset TilePreset { get { return GetPreset(EditorPrefs.GetString("scenario/tilePreset")); } }
+        public static string TeamIdKey { get 
+            {
+                if (string.IsNullOrEmpty(teamIdKey))
+                {
+                    teamIdKey = EditorPrefs.GetString("TeamIdKey");
```
Comment (Morgan-6Freedom) [inline package/Editor/Settings/PluginSettings.cs:23]: can you save the string in "scenario/teamIdKey" like the other editor prefs plz ?

**Review by Morgan-6Freedom — CHANGES_REQUESTED** [review]

File: package/Editor/Settings/PluginSettings.cs (hunk 12 above)
Comment (TheLLspectre) [reply package/Editor/Settings/PluginSettings.cs:23]: Oh ! I miss this !

File: package/Editor/PromptWindow/PromptWindow.cs (hunk 10 above)
Comment (TheLLspectre) [reply package/Editor/PromptWindow/PromptWindow.cs:1]: My visual loves to add namespace ......

File: package/Editor/Settings/PluginSettings.cs (hunk 12 above)
Comment (TheLLspectre) [reply package/Editor/Settings/PluginSettings.cs:23]: > can you save the string in "scenario/teamIdKey" like the other editor prefs plz ?

That mean I changed other one

image

File: package/Editor/PromptWindow/PromptWindowUI.cs (hunk 11 above)
Comment (Morgan-6Freedom) [reply package/Editor/PromptWindow/PromptWindowUI.cs:147]: can you change the name accordingly ? Something like "silentFetch" or "silenceMode" or something in this flavor ?

File: package/Editor/Isometric Workflow/IsometricWorkflow.cs (hunk 6 above)
Comment (TheLLspectre) [reply package/Editor/Isometric Workflow/IsometricWorkflow.cs:201]: Is inside the method to regenerate an inference 🤔

File: package/Editor/Isometric Workflow/IsometricWorkflow.cs (hunk 5 above)
Comment (TheLLspectre) [reply package/Editor/Isometric Workflow/IsometricWorkflow.cs:151]: Now it's fix in a variable and that's mean in the futur we can have a custom prompt on it ?

**Comment by TheLLspectre** [comment]
Fixing conflicts after merging dev with 125

**Review by Morgan-6Freedom — APPROVED** [review]
Seems great to me !
I would need a video that show off this whole new workflow to the team

**Comment by qvaleroo** [comment]
Select buttons on Step 1 are broken after using the Select option on Custom
https://github.com/scenario-labs/Scenario-Unity/assets/95273550/a7bd1bbc-a93e-40d7-ac69-f65598e58e35

**Comment by qvaleroo** [comment]
Isn't possible to go back to Previous Step?
Display of Step2 is not really optimized either. Window cut the content

image

**Comment by qvaleroo** [comment]
lora1, lora2.... are final names?
image

**Comment by qvaleroo** [comment]
Is it possible to click on the whole image rather than the little checkbox?
image

**Comment by TheLLspectre** [comment]
> Step1 window is not correctly sized when opening the first time on my mac: image
> 
> And is it possible to handle errors if models are not available rather than fetch error on console? Maybe we can have an outage or a model change, and we can just display an error message readable for the user, like "Model xxxx not available"?

You try to access it whitout the isometric workspace @qvaleroo  ?

**Comment by TheLLspectre** [comment]
> Select button display on Custom is broken: image

Thanks @qvaleroo  ! I didn't see that !

**Comment by qvaleroo** [comment]
I didn't get why after step4 the prompt window automatically opened, or maybe just get in front row if it was open in background 🤔 
image

**Comment by qvaleroo** [comment]
All titles change colour when I hover my mouse on it, we should avoid this as it is just a title (maybe it is only on Mac?)
image

**Comment by qvaleroo** [comment]
Regarding step5:
- You can click on buttons when they should be unclickable.
- Restart 4 generations triggers an exception

https://github.com/scenario-labs/Scenario-Unity/assets/95273550/b6c3153f-146f-4791-b680-9c35270f9ede

**Comment by TheLLspectre** [comment]
> lora1, lora2.... are final names? image

Some models will be removed

**Comment by TheLLspectre** [comment]
> Select buttons on Step 1 are broken after using the Select option on Custom https://github.com/scenario-labs/Scenario-Unity/assets/95273550/a7bd1bbc-a93e-40d7-ac69-f65598e58e35

Not looking broken, if custom image are not filled you can't select it. The select button had a trouble. And same for square if you want to select it the selector goes back to null. May be it's not clear

**Comment by qvaleroo** [comment]
> > Select buttons on Step 1 are broken after using the Select option on Custom https://github.com/scenario-labs/Scenario-Unity/assets/95273550/a7bd1bbc-a93e-40d7-ac69-f65598e58e35
> 
> Not looking broken, if custom image are not filled you can't select it. The select button had a trouble. And same for square if you want to select it the selector goes back to null. May be it's not clear

In this case we should not be able to click on the Custom checkbox. Currently you can click and see an animation. Checkbox should just be disabled. Then if you click again on a selected checkbox, do not select another one. please leave everything unselect.

**Comment by TheLLspectre** [comment]
> Step 4 is really hard to understand, I didn't get I had to enter name of building in the field
> Screen.Recording.2024-03-26.at.14.07.40.mov

Yeap, I'm agree, I add a subtitle to add indication... but if you have any idea to improve it ????

**Comment by qvaleroo** [comment]
> > Step 4 is really hard to understand, I didn't get I had to enter name of building in the field
> > Screen.Recording.2024-03-26.at.14.07.40.mov
> 
> Yeap, I'm agree, I add a subtitle to add indication... but if you have any idea to improve it ????

Maybe we can add buttons above like the prompt builder o nthe web to quickly add building types and understand what it is ? The Add Samples button is little bit mysterious and we don't know what will happen when clicking on it
image

**Comment by Morgan-6Freedom** [comment]
@qvaleroo which models do we keep for the release in this window ?
Could you make sure they will be available publicly before we merge into develop ?
image

File: package/Editor/Isometric Workflow/IsometricWorkflow.cs [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+        {
+            isVisible = true;
+        }
+
+        private void OnBecameInvisible()
+        {
+            isVisible = false;
+        }
+
+        public enum ModelStyle
```
Comment (Morgan-6Freedom) [inline package/Editor/Isometric Workflow/IsometricWorkflow.cs:324]: This is the part we need to check with @qvaleroo

**Review by Morgan-6Freedom — APPROVED** [review]

**Comment by qvaleroo** [comment]
We should not need the user to fill Workspace ID in settings. If you want the ID, you can have it with /teams endpoint
image

**Comment by qvaleroo** [comment]
Maybe rename the settings modal:

image

**Comment by qvaleroo** [comment]
Is it possible to center the text and icon for custom isometric choice?
image

**Comment by qvaleroo** [comment]
Some icons are empty
image

**Comment by qvaleroo** [comment]
image
Maybe replace "This consume Credits" by the actual cost value. You can obtain it sending the request with parameter dryRun = true (check with frontend team or backend team if help is needed)

**Comment by qvaleroo** [comment]
We should also handle Parallel inferences limit. Creator plan is limited to 5 inferences in // for example but I can send more than 5 ifnerences.
You should queue the request when receiving a 429 error, and relaunch it once a request is finished for example.
You have the max number of allowed parallel inferences using /team/{teamId}/limits endpoint

image

**Comment by TheLLspectre** [comment]
> Some icons are empty image

Some models are not available on download, when you're not on isometric workspace :/

**Review by qvaleroo — APPROVED** [review]

**Pull request merged**